crates/sui-core/src/unit_tests/move_package_tests.rs

```diff
@@ -0,0 +1,169 @@
+    let (std_move_pkg, sui_move_pkg) = sui_framework::make_std_sui_move_pkgs();
+    let modules = build_test_modules("object_basics");
+    let move_pkg = MovePackage::new_initial(
+        OBJECT_START_VERSION,
+        modules,
+        u64::MAX,
+        [&std_move_pkg, &sui_move_pkg],
+    )
+    .unwrap();
+
+    let mod_name = "object_basics";
+    let pkg_id = move_pkg.id();
+    let type_origin_table = move_pkg.type_origin_table();
+    // this package defines 4 structs: Name, NewValueEvent, Object, and Wrapper
+    assert!(type_origin_table.len() == 4);
+    verify_object_basics_initial_types(type_origin_table, &pkg_id, mod_name).await;
+
+    let linkage_table = move_pkg.linkage_table();
+    verify_object_basics_linkage(linkage_table).await;
+}
+
+#[tokio::test]
+async fn test_simple_move_upgraded_package_create() {
+    let (std_move_pkg, sui_move_pkg) = sui_framework::make_std_sui_move_pkgs();
+    let modules = build_test_modules("object_basics");
+    let move_pkg = MovePackage::new_initial(
+        OBJECT_START_VERSION,
+        modules,
+        u64::MAX,
+        [&std_move_pkg, &sui_move_pkg],
+    )
+    .unwrap();
+
+    let upgraded_modules = build_test_modules("object_basics_upgraded");
+    let storage_id = ObjectID::random();
+    let upgraded_move_pkg = move_pkg
+        .new_upgraded(
+            storage_id,
+            upgraded_modules,
+            u64::MAX,
+            [&std_move_pkg, &sui_move_pkg],
+        )
+        .unwrap();
+
+    let mod_name = "object_basics";
+    let pkg_id = move_pkg.id();
+    let type_origin_table = move_pkg.type_origin_table();
+    // the original version of this package defines 4 structs: Name, NewValueEvent, Object, and Wrapper
+    assert!(type_origin_table.len() == 4);
+    verify_object_basics_initial_types(type_origin_table, &pkg_id, mod_name).await;
+    let linkage_table = move_pkg.linkage_table();
+    verify_object_basics_linkage(linkage_table).await;
+
+    let upgraded_type_origin_table = upgraded_move_pkg.type_origin_table();
+    // the upgraded version of this package defines 5 structs: Name, NewValueEvent, Object, and
+    // Wrapper (from the original version) and ObjectNewVersion (from the upgraded version)
+
+    // check entries from the original package
+    verify_object_basics_initial_types(upgraded_type_origin_table, &pkg_id, mod_name).await;
+    // check the entry added by the upgraded package
+    assert!(
+        upgraded_type_origin_table
+            .get(&ModuleStruct {
+                module_name: mod_name.to_string(),
+                struct_name: "ObjectNewVersion".to_string(),
+            })
+            .unwrap()
+            == &storage_id
+    );
+
+    let upgraded_linkage_table = move_pkg.linkage_table();
+    // the linkage table in this upgraded package is actually the same as in the original version as
+    // dependencies haven't changed
+    verify_object_basics_linkage(upgraded_linkage_table).await;
```

Comment:
Can we manufacture a case where the linkage table does change? e.g. I believe there is a `depends_on_basics` package and an upgraded `object_basics` floating around now.